mailer/engine.py

```diff
@@ -158,24 +181,17 @@ def send_all():
                     email.connection = None
                     message.email = email  # For the sake of MessageLog
                     MessageLog.objects.log(message, RESULT_SUCCESS)
-                    sent += 1
+                    counts['sent'] += 1
                 else:
                     logging.warning("message discarded due to failure in converting from DB. Added on '%s' with priority '%s'" % (message.when_added, message.priority))  # noqa
                 message.delete()
 
-            except (socket_error, smtplib.SMTPSenderRefused,
-                    smtplib.SMTPRecipientsRefused,
-                    smtplib.SMTPDataError,
-                    smtplib.SMTPAuthenticationError) as err:
-                message.defer()
-                logging.info("message deferred due to failure: %s" % err)
-                MessageLog.objects.log(message, RESULT_FAILURE, log_message=str(err))
-                deferred += 1
-                # Get new connection, it case the connection itself has an error.
-                connection = None
+            except Exception as err:
+                connection, action = ERROR_HANDLER(connection, message, err)
+                counts[action] = counts[action] + 1
```

Comment:
Cool idea!

I would go even further and limit the try-except block to `email.send()`. What do you think?